failsafegrpc/client.go

```diff
@@ -0,0 +1,84 @@
+package failsafegrpc
+
+import (
+	"context"
+	"github.com/failsafe-go/failsafe-go"
+	"google.golang.org/grpc"
+)
+
+// UnaryClientResponse is a struct that contains the context, method, request, and response of a unary gRPC call.
+type UnaryClientResponse struct {
+	Ctx    context.Context
+	Method string
+	Req    any
+	Reply  any
+}
+
+// UnaryClientInterceptor returns a gRPC unary client interceptor that wraps the invoker with a failsafe executor.
+func UnaryClientInterceptor(executor failsafe.Executor[*UnaryClientResponse]) grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req, reply any, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		executorWithContext := executor.WithContext(ctx)
+
+		operation := func(exec failsafe.Execution[*UnaryClientResponse]) (*UnaryClientResponse, error) {
+			err := invoker(ctx, method, req, reply, cc, opts...)
+
+			resp := &UnaryClientResponse{
+				Ctx:    ctx,
+				Method: method,
+				Req:    req,
+				Reply:  reply,
+			}
+			if err != nil {
+				return resp, err
+			}
+
+			return resp, nil
+		}
+
+		_, err := executorWithContext.GetWithExecution(operation)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+}
+
+// StreamClientResponse is a struct that contains the context, stream descriptor, method, and streamer of a stream gRPC call.
+type StreamClientResponse struct {
+	Ctx    context.Context
+	Desc   *grpc.StreamDesc
+	Method string
+	Stream grpc.ClientStream
+}
+
+// StreamClientInterceptor returns a gRPC stream client interceptor that wraps the streamer with a failsafe executor.
+// If you want to use the response of the gRPC call in the policies, override `RecvMsg` and `SendMsg` in the grpc.ClientStream.
+func StreamClientInterceptor(executor failsafe.Executor[*StreamClientResponse]) grpc.StreamClientInterceptor {
+	return func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+		executorWithContext := executor.WithContext(ctx)
+
+		operation := func(exec failsafe.Execution[*StreamClientResponse]) (*StreamClientResponse, error) {
+			clientStream, err := streamer(ctx, desc, cc, method, opts...)
+
+			resp := &StreamClientResponse{
+				Ctx:    ctx,
+				Desc:   desc,
+				Method: method,
+				Stream: clientStream,
+			}
+			if err != nil {
+				return resp, err
+			}
+
+			return resp, nil
+		}
+
+		result, err := executorWithContext.GetWithExecution(operation)
```

Comment:
It looks like `StreamClientInterceptor` is very different from `UnaryClientInterceptor`. While `UnaryClientInterceptor` performs an actual RPC call and failsafe-go will handle any errors, `StreamClientInterceptor` doesn't actually perform an RPC call, instead it just creates a `ClientInterceptor`. So what we're doing here is just handling any errors with the `ClientInterceptor` creation. Is that what we want? Should we be intercepting `ClientStream.RecvMsg` calls instead, and using an executor there? 

If we did intercept and handle errors in `RecvMsg`, should we re-create the ClientInterceptor internally before doing retries? I'm not sure what the right thing to do here is.

Reply:
I noticed this client stream retry implementation https://github.com/grpc-ecosystem/go-grpc-middleware/blob/main/interceptors/retry/retry.go#L140 captures messages that are sent through the client stream, creates a new client stream before retrying, and resends the captured messages. I'm not sure how I feel about this, especially if a stream is long lived and may contain a lot of communication back and forth.

In general - retrying stream messages might be tricky for the receiver code, since it might receive the same response multiple times.

Reply:
What do you think of moving the streaming code to a separate PR so I can just merge the unary stuff for now? The complexity I mentioned above just exists for streaming calls.

Reply:
> So what we're doing here is just handling any errors with the ClientInterceptor creation. Is that what we want?

Not we want. I saw code which you mention. there are a lot of consisering.

> What do you think of moving the streaming code to a separate PR so I can just merge the unary stuff for now?

Yes. I agree.

Reply:
[0536f97](https://github.com/failsafe-go/failsafe-go/pull/54/commits/0536f97831cf792a8b89fd58327a68cfa82b1f76)